shell/shell.js

```diff
@@ -16,22 +16,153 @@
 
 const { classes: Cc, interfaces: Ci, results: Cr, utils: Cu } = Components;
 const { Runtime } = Cu.import('resource://qbrt/modules/Runtime.jsm', {});
+const { Services } = Cu.import('resource://gre/modules/Services.jsm', {});
 
-window.addEventListener('load', event => {
-  const browser = document.getElementById('content');
-  const url = window.arguments[0];
+let Accelerators = {
+  isMac: Services.appinfo.OS === 'Darwin',
+  keys: {
+    i: 73,
+    r: 82,
+    f5: 116,
+  },
+  Reload: function(event) {
+    const f5 = event.keyCode === this.keys.f5;
+    if (f5) {
+      return true;
+    }
+    if (this.isMac) {
+      // `Cmd + R`.
+      const cmdR = event.metaKey && event.keyCode === this.keys.r;
+      return cmdR;
+    }
+    // `Ctrl + Shift + R`.
+    const ctrlShiftR = event.ctrlKey && event.keyCode === this.keys.r;
+    return ctrlShiftR;
+  },
+  HardReload: function(event) {
+    // `Shift + F5`.
+    const shiftF5 = event.shiftKey && event.keyCode === this.keys.f5;
+    if (shiftF5) {
+      return true;
+    }
+    if (this.isMac) {
+      // `Cmd + Shift + R`.
+      const cmdShiftR = event.metaKey && event.shiftKey &&
+        event.keyCode === this.keys.r;
+      return cmdShiftR;
+    }
+    // `Ctrl + Shift + R`.
+    const ctrlShiftR = event.ctrlKey && event.shiftKey &&
+      event.keyCode === this.keys.r;
+    return ctrlShiftR;
+  },
+  ToggleDevTools: function(event) {
+    if (this.isMac) {
+      // `Cmd + Alt + I`.
+      const cmdAltI = event.metaKey && event.altKey &&
+        event.keyCode === this.keys.i;
+      return cmdAltI;
+    }
+    // `Ctrl + Shift + I`.
+    const ctrlShiftI = event.ctrlKey && event.shiftKey &&
+      event.keyCode === this.keys.i;
+    return ctrlShiftI;
+  },
+};
+
+function Shortcuts() {
+  this.accelerators = Accelerators;
+  this.registered = [];
+}
+Shortcuts.prototype.register = function(acceleratorName, callback) {
+  this.registered.push([acceleratorName, callback]);
+};
+Shortcuts.prototype.testEvent = function(event, acceleratorName) {
+  if (!(acceleratorName in this.accelerators)) {
+    return false;
+  }
+  return this.accelerators[acceleratorName](event);
+};
+Shortcuts.prototype.handleEvent = function(event) {
+  for (let idx = 0; idx < this.registered.length; idx++) {
+    let [ acceleratorName, callback ] = this.registered[idx];
+    if (this.testEvent(event, acceleratorName)) {
+      callback();
+    }
+  }
+};
+
+window.addEventListener('load', () => {
+  UI.init();
+}, { once: true });
```

Comment:
Nit: since the event listener only calls *UI.init*, it'd be simpler to pass that function to *window.addEventListener* directly (binding it to the UI object to ensure that its context is as expected):

```
window.addEventListener('load', UI.init.bind(UI), { once: true });
```

(If you did that, then the _window.addEventListener_ call would need to move below the _UI_ definition, so _UI.init_ is defined by the time it gets bound and added.)